controllers/operator/mongodbreplicaset_controller.go

```diff
@@ -238,54 +311,94 @@ func (r *ReconcileMongoDbReplicaSet) Reconcile(ctx context.Context, request reco
-	lastSpec, err := rs.GetLastSpec()
-	if err != nil {
-		lastSpec = &mdbv1.MongoDbSpec{}
-	}
-
-	publishAutomationConfigFirst := publishAutomationConfigFirstRS(ctx, r.client, *rs, lastSpec, deploymentOpts.currentAgentAuthMode, projectConfig.SSLMMSCAConfigMap, log)
+	publishAutomationConfigFirst := publishAutomationConfigFirstRS(ctx, reconciler.client, *rs, r.deploymentState.LastAchievedSpec, deploymentOpts.currentAgentAuthMode, projectConfig.SSLMMSCAConfigMap, log)
 	status := workflow.RunInGivenOrder(publishAutomationConfigFirst,
 		func() workflow.Status {
-			return r.updateOmDeploymentRs(ctx, conn, rs.Status.Members, rs, log, tlsCertPath, internalClusterCertPath, deploymentOpts, shouldMirrorKeyfileForMongot, false).OnErrorPrepend("Failed to create/update (Ops Manager reconciliation phase):")
+			return r.updateOmDeploymentRs(ctx, conn, r.deploymentState.LastReconcileMemberCount, tlsCertPath, internalClusterCertPath, deploymentOpts, shouldMirrorKeyfileForMongot, false).OnErrorPrepend("failed to create/update (Ops Manager reconciliation phase):")
 		},
 		func() workflow.Status {
-			return r.reconcileMemberResources(ctx, rs, conn, log, projectConfig, deploymentOpts)
+			return r.reconcileMemberResources(ctx, conn, projectConfig, deploymentOpts)
 		})
 
 	if !status.IsOK() {
-		return r.updateStatus(ctx, rs, status, log)
+		return r.updateStatus(ctx, status)
 	}
 
 	// === 6. Final steps
 	if scale.IsStillScaling(rs) {
-		return r.updateStatus(ctx, rs, workflow.Pending("Continuing scaling operation for ReplicaSet %s, desiredMembers=%d, currentMembers=%d", rs.ObjectKey(), rs.DesiredReplicas(), scale.ReplicasThisReconciliation(rs)), log, mdbstatus.MembersOption(rs))
+		return r.updateStatus(ctx, workflow.Pending("Continuing scaling operation for ReplicaSet %s, desiredMembers=%d, currentMembers=%d", rs.ObjectKey(), rs.DesiredReplicas(), scale.ReplicasThisReconciliation(rs)), mdbstatus.MembersOption(rs))
 	}
 
-	annotationsToAdd, err := getAnnotationsForResource(rs)
-	if err != nil {
-		return r.updateStatus(ctx, rs, workflow.Failed(err), log)
+	// Write state and vault annotations on successful reconciliation
+	if err := r.writeState(ctx); err != nil {
+		return r.updateStatus(ctx, workflow.Failed(xerrors.Errorf("failed to write state: %w", err)))
 	}
 
-	if vault.IsVaultSecretBackend() {
-		secrets := rs.GetSecretsMountedIntoDBPod()
-		vaultMap := make(map[string]string)
-		for _, s := range secrets {
-			path := fmt.Sprintf("%s/%s/%s", r.VaultClient.DatabaseSecretMetadataPath(), rs.Namespace, s)
-			vaultMap = merge.StringToStringMap(vaultMap, r.VaultClient.GetSecretAnnotation(path))
-		}
-		path := fmt.Sprintf("%s/%s/%s", r.VaultClient.OperatorScretMetadataPath(), rs.Namespace, rs.Spec.Credentials)
-		vaultMap = merge.StringToStringMap(vaultMap, r.VaultClient.GetSecretAnnotation(path))
-		for k, val := range vaultMap {
-			annotationsToAdd[k] = val
+	if err := r.writeVaultAnnotations(ctx); err != nil {
```

Comment:
Both `writeState` and `writeVaultAnnotations` call `annotations.SetAnnotations` which gets the object and patches it.

If we can reduce write operations to the API server - that would be great.

Reply:
I fixed this and we write only once now